lib/ansible/plugins/inventory/toml.py

```diff
@@ -0,0 +1,240 @@
+# Copyright (c) 2018 Matt Martz <matt@sivel.net>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    inventory: toml
+    version_added: "2.8"
+    short_description: Uses a specific TOML file as an inventory source.
+    description:
+        - TOML based inventory format
+        - File MUST have a valid '.toml' file extension
+    notes:
+        - To function it requires the 'toml' plugin being whitelisted in configuration.
+        - Requires the 'toml' python library
+'''
+
+EXAMPLES = '''
+example1: |
+    [all.vars]
+    has_java = false
+
+    [web]
+    children = [
+        "apache",
+        "nginx"
+    ]
+    vars = { http_port = 8080, myvar = 23 }
+
+    [web.hosts]
+    host1 = {}
+    host2 = { ansible_port = 222 }
+
+    [apache.hosts]
+    tomcat1 = {}
+    tomcat2 = { myvar = 34 }
+    tomcat3 = { mysecret = "03#pa33w0rd" }
+
+    [nginx.hosts]
+    jenkins1 = {}
+
+    [nginx.vars]
+    has_java = true
+
+example2: |
+    [all.vars]
+    has_java = false
+
+    [web]
+    children = [
+        "apache",
+        "nginx"
+    ]
+
+    [web.vars]
+    http_port = 8080
+    myvar = 23
+
+    [web.hosts.host1]
+    [web.hosts.host2]
+    ansible_port = 222
+
+    [apache.hosts.tomcat1]
+
+    [apache.hosts.tomcat2]
+    myvar = 34
+
+    [apache.hosts.tomcat3]
+    mysecret = "03#pa33w0rd"
+
+    [nginx.hosts.jenkins1]
+
+    [nginx.vars]
+    has_java = true
+
+example3: |
+    [ungrouped.hosts]
+    host1 = {}
+    host2 = { ansible_host = "127.0.0.1", ansible_port = 44 }
+    host3 = { ansible_host = "127.0.0.1", ansible_port = 45 }
+
+    [g1.hosts]
+    host4 = {}
+
+    [g2.hosts]
+    host4 = {}
+'''
+
+import os
+
+from collections import MutableMapping, MutableSequence
+
+from ansible.errors import AnsibleFileNotFound, AnsibleParserError
+from ansible.module_utils.six import string_types, text_type
+from ansible.module_utils._text import to_bytes, to_native
+from ansible.parsing.yaml.objects import AnsibleSequence, AnsibleUnicode
+from ansible.plugins.inventory import BaseFileInventoryPlugin
+
+try:
+    import toml
+    HAS_TOML = True
+except (ImportError, AttributeError):
+    HAS_TOML = False
+
+
+if HAS_TOML and hasattr(toml, 'TomlEncoder'):
+    class AnsibleTomlEncoder(toml.TomlEncoder):
+        def __init__(self, *args, **kwargs):
+            super(AnsibleTomlEncoder, self).__init__(*args, **kwargs)
+            self.dump_funcs.update({
+                AnsibleSequence: self.dump_funcs.get(list),
+                AnsibleUnicode: self.dump_funcs.get(str),
```

Comment:
`str` -> `six.text_type` I guess?

cc @abadger ^

Reply:
Not here.  This is updating `self.dump_funcs` which is a `toml` internal thing.  I need to get the "dump_func" that `toml` has listed for `str`.

Basically to make `AnsibleUnicode` behave the same as `toml` specifies for `str`.

Reply:
But that's bytes under py2. Wouldn't it cause problems?

Reply:
`toml` has no dumper at all for `bytes`.  And I haven't run into a situation that has necessitated it.

Additionally `toml` uses the same dumper for `str` and `unicode`: 

https://github.com/uiri/toml/blob/f849546caac716cee7f2f32e1b1fe7e2822e342a/toml/encoder.py#L115-L125

I mean, I suppose I could try to do something tricky here, but I don't see need.